blockchain/types/derive_sha.go

```diff
@@ -40,15 +41,32 @@ const (
 )
 
 var (
-	// EmptyRootHashOriginal is the empty root hash of a state trie,
-	// which is equal to EmptyRootHash with ImplDeriveShaOriginal.
-	EmptyRootHashOriginal = common.HexToHash("0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421")
+	commonEmptyRootHash = common.HexToHash("0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421")
+
+	// Root hash of an empty trie.
+	// Used to represent a block's empty StateRoot or an account's StorageRoot.
+	EmptyRootHash = commonEmptyRootHash
+
+	// Hash of an empty byte array.
+	// Used to represent an account's empty contract code.
+	EmptyCodeHash = crypto.Keccak256Hash(nil) // c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470
+
+	// Empty TransactionRoot and ReceiptsRoot for the Original algorithm.
+	EmptyTxRootOriginal = EmptyRootHash
+
+	// Empty TransactionRoot and ReceiptsRoot for the Simple algorithm.
+	EmptyTxRootSimple = EmptyCodeHash
+
+	// Empty TransactionRoot and ReceiptsRoot for the Concat algorithm.
+	EmptyTxRootConcat = EmptyCodeHash
```

Comment:
How about using `EmptyDeriveShaOrig`, `EmptyDeriveShaSimple`, and `EmptyDeriveShaConcat` for clarity?

Reply:
I'd prefer TxRoot. I did think about TxReceiptRoot or TxAndReceiptRoot but both are wordy. On the other hand, DeriveSha sounds too generic to me.

Reply:
Ok, then let's keep the names. Thanks.